### DISCUSSION TITLE
api rules for filtering through multiple relation fields

### DISCUSSION THREAD

**Discussion by k4io** [pocketbase/pocketbase#5560] [Q&A] [answered]
hi i am trying to setup permissions and roles collections (shortened collection defs for readability)

roles:
```
{
    "name": "user_roles",
    "type": "base",
    "schema": [
      {
        "system": false,
        "id": "zcpkgf3s",
        "name": "name",
        "type": "text",
      },
      {
        "system": false,
        "id": "tkcmusqy",
        "name": "permissions",
        "type": "relation",
        "unique": false,
      }
    ]
  }
```
(only has name, permissions)

permissions:

```
{
    "name": "user_permissions",
    "schema": [
      {
        "system": false,
        "id": "zgsugnxm",
        "name": "panel",
        "type": "relation",
      },
      {
        "system": false,
        "id": "lti3gsrh",
        "name": "resource",
        "type": "text",
      },
      {
        "system": false,
        "id": "xadb2aoe",
        "name": "action",
        "type": "text",
      },
      {
        "system": false,
        "id": "imroc6nl",
        "name": "conditions",
        "type": "text",
      },
    ]
  }
```

(has fields: panel (single relation), resource (text), action (text), conditions (text))

I am trying to setup the permissions with the rules API like this:

```
@request.auth.id != "" && @request.auth.account_enabled = true && 
(
  panel = @request.query.panel && 
  @request.auth.role.permissions.panel.id = @request.query.panel
)
```

called like:

```
let records = await this.pb.collection('records').getFullList({
      sort: '-created',
      panel: panel.id,
    });
```

I have the correct role on my logged-in user and that role has the correct permissions but it **does not** return _any_ data.

I have also tried using :each like:

```
@request.auth.id != "" && @request.auth.account_enabled = true && 
(
  panel = @request.query.panel && 
  @request.auth.role.permissions.panel.id:each = @request.query.panel
)
```
and
```
@request.auth.id != "" && @request.auth.account_enabled = true && 
(
  panel = @request.query.panel && 
  @request.auth.role.permissions:each.panel.id = @request.query.panel
)
```

but both to no avail. Does anyone have any ideas on a solution for this?

Permissions setup:
role -> name (text), permissions (multiple permission relation)
permission -> panel (single panel relation), resource (text), action (text), conditions (text)

**Comment by k4io** [comment, marked as answer]
I am using crud hooks instead

**Reply by ganigeorgiev** [reply]
Multi-valued fields need to use the `?` operator prefix to target only one of the compared value (aka. `Any/at-least-one-of`), otherwise it will result in "match-all" constraint. Or in other words, your initial rule should work if you replace it with `@request.auth.role.permissions.panel ?= @request.query.panel`.

I don't exactly understand what actually is the purpose of the rule in your case and extra query parameter but if you want to compare the `panel` field against the one from current logged in user, then you should be able to simply use:

```js
panel ?= @request.auth.role.permissions.panel
```
